src/v8n.js

```diff
@@ -221,9 +263,75 @@ const core = {
         throw new ValidationException(rule, value, ex);
       }
     });
+  },
+
+  /**
+   * Performs asynchronous boolean based validation.
+   *
+   * When this function is used it performs the validation process
+   * asynchronously, and it returns a promise that resolves to `true` when the
+   * validated value is valid and to `false` when the validated value is
+   * invalid.
+   *
+   * > To learn more about asynchronous validation, look at the
+   * > [Validation](#Validation) section.
+   *
+   * > For a validation strategy with non promise-based rules, you'd better use
+   * > the [test](#test) function.
+   *
+   * @param {any} value the value to be validated
+   * @returns {Promise} promise that resolves to a boolean value
+   */
+  asyncTest(value) {
+    return executeAsyncRules(value, this.chain).then(() => true, () => false);
+  },
+
+  /**
+   * Performs asynchronous failed-rule based validation.
+   *
+   * When this function is used it performs the validation process
+   * asynchronously, and it returns a promise that resolves to nothing when the
+   * validated value is valid and rejects with the first failed
+   * {@link Rule rule} when it's invalid.
+   *
+   * > To learn more about asynchronous validation, look at the
+   * > [Validation](#Validation) section.
+   *
+   * > For a validation strategy with non promise-based rules, you'd better use
+   * > the [check](#check) function.
+   *
+   * @param {any} value the value to be validated
+   * @returns {Promise} promise that rejects with the first-failed
+   * {@link Rule rule} when value is invalid
+   */
+  asyncCheck(value) {
+    return executeAsyncRules(value, this.chain);
   }
 };
 
+function executeAsyncRules(value, rules) {
+  return new Promise((resolve, reject) => {
+    executeAsyncRulesAux(value, rules, 0, resolve, reject);
+  });
+}
+
+function executeAsyncRulesAux(value, rules, index, resolve, reject) {
+  if (index < rules.length) {
+    const rule = rules[index];
+    const result = Promise.resolve(rule.fn(value));
+
+    result.then(valid => {
+      if (valid !== rule.invert) {
+        executeAsyncRulesAux(value, rules, index + 1, resolve, reject);
```

Comment:
Pass shifted rules here instead of the entire array and the index.

Reply:
Nice, that's a good improvement for clean code.